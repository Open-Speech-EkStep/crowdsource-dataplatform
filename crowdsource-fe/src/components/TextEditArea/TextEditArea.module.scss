.addText {
  box-shadow: inset 0 0 16px rgba(15, 39, 73, 0.4);
}

.addTextDisabled {
  box-shadow: inset 0 0 18px #d8d8d8;
}

.addTextError {
  box-shadow: inset 0 0 16px rgba(211, 47, 47, 0.4);
}

.textarea {
  height: 6.25rem;
  resize: none;

  &:disabled {
    background-color: var(--light-color);
  }

  &:focus {
    box-shadow: none;
  }
}

.keyboard {
  background-color: #f1f1f1;
  border: 1px solid #d3d3d3;
  padding: 0.8rem 10px;
  position: absolute;
  width: 600px;
  z-index: 1;
}

.crossIcon {
  cursor: pointer;
  position: absolute;
  right: 15px;
}

.keyboardIcon {
  bottom: 0.5rem;
  opacity: 0.8;
  right: 0.75rem;

  button {
    background-color: transparent;
  }

  &:hover {
    opacity: 1;
  }
}

.textareaWrapper,
.textareaLabel {
  opacity: 0.63;
}

.textAreaDisabled {
  opacity: 0.28;
}
