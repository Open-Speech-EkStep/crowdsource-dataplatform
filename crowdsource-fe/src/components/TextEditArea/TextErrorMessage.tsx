interface TextErrorMessageProps {
  message: string;
}

const TextErrorMessage = ({ message }: TextErrorMessageProps) => {
  return (
    <span data-testid="ErrorText" className="d-block text-danger fst-italic mt-2 display-5">
      {message}
    </span>
  );
};

export default TextErrorMessage;
