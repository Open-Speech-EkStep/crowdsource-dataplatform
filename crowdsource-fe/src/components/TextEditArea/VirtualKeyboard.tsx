import { useRef } from 'react';

import { useTranslation } from 'next-i18next';
import Draggable from 'react-draggable';
import Keyboard from 'react-simple-keyboard';

import styles from './TextEditArea.module.scss';

interface VirtualKeyboardProps {
  language: string;
  keyBoardRef: Function;
  setShowKeyboard: Function;
  onChange: Function;
  onKeyPress: Function;
  layoutName: string;
  layout: Record<string, Array<string>>;
}

const VirtualKeyboard = ({
  language,
  setShowKeyboard,
  onChange,
  onKeyPress,
  layoutName,
  layout,
  keyBoardRef,
}: VirtualKeyboardProps) => {
  const { t } = useTranslation();

  const nodeRef = useRef(null);

  return (
    <Draggable bounds="main" nodeRef={nodeRef}>
      <div ref={nodeRef} data-testid="virtual-keyboard" className={`${styles.keyboard} d-none d-md-block`}>
        <div className="pb-2">
          <span>{t(language.toLowerCase())}</span>
          <span
            data-testid="close-keyboard"
            className={styles.crossIcon}
            onClick={() => {
              setShowKeyboard(false);
            }}
            aria-hidden="true"
          >
            X
          </span>
        </div>

        <Keyboard
          keyboardRef={keyBoardRef}
          onChange={onChange}
          onKeyPress={onKeyPress}
          layoutName={layoutName}
          layout={layout}
        />
      </div>
    </Draggable>
  );
};

export default VirtualKeyboard;
