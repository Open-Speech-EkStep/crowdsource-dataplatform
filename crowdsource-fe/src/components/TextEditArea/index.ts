export { default } from './TextEditArea';
export { default as ErrorText } from './TextErrorMessage';
