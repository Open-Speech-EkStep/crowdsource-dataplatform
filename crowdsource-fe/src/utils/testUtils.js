import { render } from '@testing-library/react';

const Wrapper = ({ children }) => children;

const customRender = (ui, options) => render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };
