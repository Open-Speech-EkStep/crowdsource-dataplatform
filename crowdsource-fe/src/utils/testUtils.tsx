/* eslint-disable jest/no-export */
import type { ReactNode, ReactElement, ComponentType } from 'react';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import puppeteer from 'puppeteer';
import type { Browser } from 'puppeteer';

const Wrapper: ComponentType = ({ children }: { children?: ReactNode }) => children as ReactElement;

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };

export const verifyAxeTest = ({ container }: { container: RenderResult['container'] }) => {
  it('should not fail an axe audit', async () => {
    let axeResult;

    await act(async () => {
      axeResult = await axe(container);
    });

    expect(axeResult).toHaveNoViolations();
  });
};

export const verifyVRTest = (title: string, url: string) => {
  describe(`${title} VR test`, () => {
    let browser: Browser;

    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: true,
        devtools: false,
        ignoreHTTPSErrors: true,
        args: ['--start-fullscreen'],
      });
    });

    it('match snapshot', async () => {
      const page = await browser.newPage();
      await page.setViewport({ width: 1920, height: 1080 });

      await page.goto(url, {
        waitUntil: 'networkidle2',
      });

      await page.addStyleTag({
        content: `
      *,
      *::after,
      *::before {
          transition-delay: 0s !important;
          transition-duration: 0s !important;
          animation-delay: -0.0001s !important;
          animation-duration: 0s !important;
          animation-play-state: paused !important;
          caret-color: transparent !important;
      }
      `,
      });

      const image = await page.screenshot({
        fullPage: true,
      });

      expect(image).toMatchImageSnapshot();
    });

    afterAll(async () => {
      await browser.close();
    });
  });
};
