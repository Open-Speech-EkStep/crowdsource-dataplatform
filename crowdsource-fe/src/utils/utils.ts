import { jsPDF } from 'jspdf';
import { i18n } from 'next-i18next';

import { ErrorStatusCode } from 'constants/errorStatusCode';
import { INITIATIVES_MAPPING } from 'constants/initiativeConstants';
import { KEYBOARD_ERROR, LANGUAGE_UNICODE, OTHER_LANGUAGE_UNICODE } from 'constants/Keyboard';
import type { InitiativeType } from 'types/InitiativeType';

import apiPaths from '../constants/apiPaths';

var platform = require('platform');

export const convertIntoHrsFormat = (data: any, isSeconds = true) => {
  const hours = Math.floor(data / 3600);
  const remainingAfterHours = data % 3600;
  const minutes = Math.floor(remainingAfterHours / 60);
  const seconds = Math.round(remainingAfterHours % 60);
  if (isSeconds) {
    return { hours, minutes, seconds };
  } else {
    return { hours, minutes };
  }
};

export const roundOffValue = (average: any, precision: number) => {
  const formattedAverage = average.toFixed(precision);
  return formattedAverage;
};

export const capitalizeFirstLetter = (strValue: string) => {
  return strValue?.charAt(0).toUpperCase() + strValue?.slice(1);
};

const translateText = (text: string) => i18n?.t(text);

export const formatTime = (hours: number, minutes = 0, seconds = 0) => {
  const hrStr = translateText('hours1');
  const minStr = translateText('minutes1');
  const secStr = translateText('seconds1');
  let result = '';
  if (hours > 0) {
    result += `${hours} ${hrStr} `;
  }
  if (minutes > 0) {
    result += `${minutes} ${minStr} `;
  }
  if (hours === 0 && minutes === 0 && seconds > 0) {
    result += `${seconds} ${secStr} `;
  }

  if (hours === 0 && minutes === 0 && seconds === 0) {
    result += `0 ${secStr} `;
  }

  return result.substr(0, result.length - 1);
};

export const convertTimeFormat = (value: any) => {
  const { hours, minutes, seconds } = convertIntoHrsFormat(Number(value) * 60 * 60);
  const data = formatTime(hours, minutes, seconds);
  return data;
};

const roundDuration = (value: number) => {
  return Math.round(value * 100) / 100;
};

export const getHoursValue = (value: any) => {
  return roundDuration(value);
};

export const getMinutesValue = (value: any) => {
  return roundDuration(value * 60);
};

export const getHoursText = (value: any) => {
  return `${getHoursValue(value)} ${i18n?.t('hours2')}`;
};

export const getMinutesText = (value: any) => {
  return `${getMinutesValue(value)} ${i18n?.t('minutes2')}`;
};

export const isSunoOrBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['asr', 'text'];
  return boloOrSuno.includes(value);
};

export const isBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['text'];
  return boloOrSuno.includes(value);
};

export const findInputError = (text: string, currentModule: string, language: string) => {
  if (!text.length) return;
  const newText = text.replace(/\s/g, ''); //read input value, and remove "space" by replace \s
  let error: any = KEYBOARD_ERROR.language;
  const isSunoModule = currentModule === INITIATIVES_MAPPING.suno;
  const specialSymbols = /[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007E\u0964-\u0965]/;
  if (currentModule === INITIATIVES_MAPPING.suno && specialSymbols.test(newText) === true) {
    return KEYBOARD_ERROR.symbol;
  }
  Object.entries(isSunoModule ? LANGUAGE_UNICODE : OTHER_LANGUAGE_UNICODE).forEach(([key, value]) => {
    if (value.test(newText)) {
      //Check Unicode to see which one is true
      if (key.toLowerCase() === language.toLowerCase()) {
        error = KEYBOARD_ERROR.noError;
      }
    }
  });
  return error;
};

export const fetchLocationInfo = async () => {
  const result = await fetch('https://www.cloudflare.com/cdn-cgi/trace')
    .then(res => res.text())
    .then(async ipAddressText => {
      const dataArray = ipAddressText.split('\n');
      let ipAddress = '';
      for (let ind in dataArray) {
        if (dataArray[ind].startsWith('ip=')) {
          ipAddress = dataArray[ind].replace('ip=', '');
          break;
        }
      }
      if (ipAddress.length !== 0) {
        const data = await fetch(`${apiPaths.locationInfo}?ip=${ipAddress}`);
        return data && data.json();
      } else {
        return new Promise((resolve, reject) => {
          reject('Ip Address not available');
        });
      }
    });

  return result;
};

export const getDeviceInfo = () => {
  const os = platform.os;
  return os.family + ' ' + os.version;
};

export const getBrowserInfo = () => {
  return platform.name + ' ' + platform.version;
};

export const downloadBadge = (
  locale: string | undefined,
  initiative: string,
  source: string,
  badgeType: string,
  badgeId: string
) => {
  const pdf = new jsPDF();
  const img = new Image();
  img.src = `/images/${locale}/badges/${locale}_${initiative}_${badgeType}_${source}.png`;
  img.crossOrigin = 'Anonymous';
  img.onload = function () {
    pdf.addImage(img, 50, 10, 105, 130);
    pdf.save(`${badgeType}-badge.pdf`);
  };

  pdf.text(`Badge Id : ${badgeId}`, 36, 190);
};

export const groupBy = function (list: any, keyName: any) {
  if (!list || list.length === 0 || !list[0][keyName]) {
    return [];
  }
  return list.reduce(function (pair: any, key: any) {
    (pair[key[keyName]] = pair[key[keyName]] || []).push(key);
    return pair;
  }, {});
};

/* istanbul ignore next */
export const visualize = (visualizer: any, analyser: any) => {
  const canvasCtx = visualizer.getContext('2d');
  const bufferLength = analyser.frequencyBinCount;
  const dataArray = new Uint8Array(bufferLength);
  const WIDTH = visualizer.width;
  const HEIGHT = visualizer.height;
  // TODO do we need to limit the number of time visualize refreshes per second
  // so that it can run on Android processors without causing audio to drop?
  function draw() {
    // this is more efficient than calling with processor.onaudioprocess
    // and sending floatarray with each call...
    requestAnimationFrame(draw);
    analyser.getByteTimeDomainData(dataArray);
    // canvasCtx.fillStyle = 'rgb(255, 255, 255, 0)';
    // canvasCtx.fillRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.clearRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.lineWidth = 2;
    canvasCtx.strokeStyle = 'rgb(0,123,255)';
    canvasCtx.beginPath();
    const sliceWidth = (WIDTH * 1.0) / bufferLength;
    let x = 0;
    for (let i = 0; i < bufferLength; i++) {
      let v = dataArray[i] / 128.0; // uint8
      let y = (v * HEIGHT) / 2; // uint8
      if (i === 0) {
        canvasCtx.moveTo(x, y);
      } else {
        canvasCtx.lineTo(x, y);
      }
      x += sliceWidth;
    }
    canvasCtx.lineTo(visualizer.width, visualizer.height / 2);
    canvasCtx.stroke();
  }

  draw();
};

export const getErrorMsg = (error: any) => {
  if (error && error.status === ErrorStatusCode.TOO_MANY_REQUEST) {
    return 'multipleRequestApiError';
  } else {
    return 'apiFailureError';
  }
};
