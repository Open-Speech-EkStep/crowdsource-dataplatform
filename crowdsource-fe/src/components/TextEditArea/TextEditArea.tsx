import React, { Fragment, useEffect, useRef, useState } from 'react';

import { useTranslation } from 'next-i18next';
import Image from 'next/image';
import Form from 'react-bootstrap/Form';
import Draggable from 'react-draggable';
import Keyboard from 'react-simple-keyboard';

import Button from 'components/Button';
import { KeyboardLanguageLayout } from 'constants/Keyboard';
import { verifyLanguage } from 'utils/utils';

import styles from './TextEditArea.module.scss';
import TextErrorMessage from './TextErrorMessage';

import 'react-simple-keyboard/build/css/index.css';

interface TextEditAreaProps {
  language: string;
  initiative: string;
  setTextValue: (value: string) => void;
  textValue?: string;
  isTextareaDisabled: boolean;
}

const TextEditArea = ({
  language,
  isTextareaDisabled,
  initiative,
  setTextValue,
  textValue,
}: TextEditAreaProps) => {
  const { t } = useTranslation();
  const [input, setInput] = useState('');
  const [showKeyboard, setShowKeyboard] = useState(false);
  const [layout, setLayout] = useState<any>();
  const [isUsingPhysicalKeyboard, setIsUsingPhysicalKeyboard] = useState(false);
  const [layoutName, setLayoutName] = useState<string>('default');
  const [showError, setShowError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');
  const nodeRef = useRef(null);
  const chromeExtension = useRef<any>(null);

  let keyboard: any;

  useEffect(() => {
    chromeExtension.current = document.getElementById('GOOGLE_INPUT_CHEXT_FLAG');
    if (!textValue) {
      setShowError(false);
    }
    setLayout(KeyboardLanguageLayout[language]);
    setInput(textValue ?? '');
  }, [language, textValue]);

  const onChange = (input: any) => {
    setIsUsingPhysicalKeyboard(false);
    setInput(input);
    const error = verifyLanguage(input, initiative, language);
    handleError(error);
  };

  const handleError = (error: any) => {
    if (error && error.type === 'language') {
      setShowError(true);
      setErrorMessage(t('typeInChosenLanguage'));
    } else if (error && error.type === 'symbol') {
      setShowError(true);
      setErrorMessage(t('specialCharacters'));
    } else {
      setTextValue(input);
      setShowError(false);
    }
  };

  const onKeyPress = (button: any) => {
    if (['{lock}', '{shift}'].includes(button)) handleShiftButton();
  };

  const handleShiftButton = () => {
    setLayoutName(layoutName === 'default' ? 'shift' : 'default');
  };

  const onChangeInput = (event: any) => {
    setIsUsingPhysicalKeyboard(true);
    const input = event.target.value;
    const error = verifyLanguage(input, initiative, language);
    handleError(error);
    setTextValue(input);
    setInput(input);
    keyboard?.setInput(input);
    if (!input) {
      setShowKeyboard(false);
    }
  };

  return (
    <Fragment>
      <div
        data-testid="TextEditArea"
        className={`
          position-relative border border-2 border-primary
          ${styles.addText} ${
          isTextareaDisabled && `${styles.addTextDisabled} border border-2 border-primary-20`
        }
          ${showError && `${styles.addTextError} border border-2 border-danger`} rounded-8  bg-light p-4
        `}
      >
        <div className={`${isTextareaDisabled && `${styles.textAreaDisabled}`}`}>
          <Form.Group controlId="textarea">
            <Form.Label className={`${styles.textareaLabel} display-6`}>
              {t('addText')} ({t(language.toLowerCase())})
            </Form.Label>
            <Form.Control
              as="textarea"
              disabled={isTextareaDisabled}
              value={input}
              onFocus={() => {
                if (!isUsingPhysicalKeyboard && !chromeExtension.current) {
                  setShowKeyboard(true);
                }
              }}
              onChange={onChangeInput}
              placeholder={t('typingPlaceholder')}
              className={`${styles.textarea} border-0 p-0 display-3`}
            />
          </Form.Group>
          <div className={`${styles.keyboardIcon} d-none  d-md-block  position-absolute`}>
            <Button variant="normal" onClick={() => setShowKeyboard(true)} className="d-flex">
              <Image src="/images/keyboard_icon.svg" width="24" height="24" alt="keyboardBtn" />
            </Button>
          </div>
        </div>
      </div>
      {showError ? <TextErrorMessage message={errorMessage} /> : null}
      {showKeyboard ? (
        <Draggable bounds="body" nodeRef={nodeRef}>
          <div
            ref={nodeRef}
            data-testid="virtual-keyboard"
            className={`${styles.keyboard} d-none d-md-block`}
          >
            <div className="pb-2">
              <span>{t(language.toLowerCase())}</span>
              <span
                data-testid="close-keyboard"
                className={styles.crossIcon}
                onClick={() => {
                  setShowKeyboard(false);
                }}
                aria-hidden="true"
              >
                X
              </span>
            </div>

            <Keyboard
              keyboardRef={(r: any) => (keyboard = r)}
              onChange={onChange}
              onKeyPress={onKeyPress}
              layoutName={layoutName}
              layout={layout}
            />
          </div>
        </Draggable>
      ) : null}
    </Fragment>
  );
};

export default TextEditArea;
