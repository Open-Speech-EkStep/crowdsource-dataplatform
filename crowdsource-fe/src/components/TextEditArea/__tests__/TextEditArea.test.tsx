import { render, verifyAxeTest, userEvent, screen, waitFor } from 'utils/testUtils';

import TextEditArea from '../TextEditArea';

describe('TextEditArea English', () => {
  const mockOnError = jest.fn();
  const setup = (contributionLanguage: string, error?: { errorTextKey: string }) =>
    render(
      <TextEditArea
        id="textarea"
        language={contributionLanguage}
        initiative="asr"
        setTextValue={value => value}
        isTextareaDisabled={false}
        label={`addText (${contributionLanguage})`}
        onError={mockOnError}
        closeKeyboard={true}
        validationError={error}
      />
    );

  verifyAxeTest(setup('English'));

  it('should render the component and matches it against stored snapshot', () => {
    const { asFragment } = setup('English');

    expect(asFragment()).toMatchSnapshot();
  });

  it('should test the textarea text with valid language', async () => {
    setup('English');

    userEvent.type(screen.getByRole('textbox', { name: 'addText (English)' }), 'abc');

    await waitFor(() => {
      expect(screen.queryByText('Please type in your chosen language')).not.toBeInTheDocument();
    });
    await waitFor(() => {
      expect(screen.queryByTestId('virtual-keyboard')).not.toBeInTheDocument();
    });
  });

  it('should show the error with "Please type in your chosen language"', async () => {
    setup('English');

    userEvent.type(screen.getByRole('textbox', { name: 'addText (English)' }), 'बपपप');

    expect(screen.getByTestId('ErrorText')).toBeInTheDocument();
    expect(screen.getByText('typeInChosenLanguage')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByTestId('virtual-keyboard')).not.toBeInTheDocument();
    });
  });

  it('should test the textarea button for setting layout from virtual keyboard', async () => {
    setup('English');
    userEvent.click(screen.getByRole('button', { name: 'keyboardBtn' }));

    await waitFor(() => {
      expect(screen.getByTestId('virtual-keyboard')).toBeInTheDocument();
    });

    userEvent.click(screen.getAllByText('shift')[1]);
    userEvent.click(screen.getAllByText('shift')[1]);
  });

  it('should test the textarea text from virtual keyboard', async () => {
    setup('English');
    userEvent.click(screen.getByRole('button', { name: 'keyboardBtn' }));

    await waitFor(() => {
      expect(screen.getByTestId('virtual-keyboard')).toBeInTheDocument();
    });

    userEvent.click(screen.getByText('caps'));
    userEvent.click(screen.getByText('A'));

    expect(screen.getByRole('textbox', { name: 'addText (English)' })).toHaveValue('A');
  });

  it('should show the virtual keyboard on button click', async () => {
    setup('English');

    userEvent.click(screen.getByRole('button', { name: 'keyboardBtn' }));

    await waitFor(() => {
      expect(screen.getByTestId('virtual-keyboard')).toBeInTheDocument();
    });

    userEvent.click(screen.getByTestId('close-keyboard'));

    await waitFor(() => {
      expect(screen.queryByText('X')).not.toBeInTheDocument();
    });
  });

  it('should hide the virtual keyboard when user empty the textfield', async () => {
    setup('English');

    userEvent.click(screen.getByRole('button', { name: 'keyboardBtn' }));

    await waitFor(() => {
      expect(screen.getByTestId('virtual-keyboard')).toBeInTheDocument();
    });

    userEvent.type(screen.getByRole('textbox', { name: 'addText (English)' }), 'abc');

    userEvent.clear(screen.getByRole('textbox', { name: 'addText (English)' }));

    await waitFor(() => {
      expect(screen.queryByText('X')).not.toBeInTheDocument();
    });
  });

  it('should not show any error message if input is empty', async () => {
    setup('English');

    expect(screen.queryByText('validationWarningText')).not.toBeInTheDocument();
  });

  it('should show validation warning message if present', async () => {
    setup('English', { errorTextKey: 'validationWarningText' });

    expect(screen.getByText('validationWarningText')).toBeInTheDocument();
  });

  it('should show ui errors on priority even if validation warning is present', async () => {
    setup('English', { errorTextKey: 'validationWarningText' });

    userEvent.clear(screen.getByRole('textbox', { name: 'addText (English)' }));
    userEvent.type(screen.getByRole('textbox', { name: 'addText (English)' }), 'abc');
    expect(screen.getByText('validationWarningText')).toBeInTheDocument();
    // Add a special chartacter in input
    userEvent.type(screen.getByRole('textbox', { name: 'addText (English)' }), 'बपपप');
    expect(screen.getByText('typeInChosenLanguage')).toBeInTheDocument();
  });
});

describe('TextEditArea Hindi', () => {
  const setup = (contributionLanguage: string) =>
    render(
      <TextEditArea
        id="textarea"
        language={contributionLanguage}
        initiative="asr"
        setTextValue={value => value}
        isTextareaDisabled={false}
        label={`addText (${contributionLanguage})`}
        onError={() => {}}
        closeKeyboard={true}
      />
    );

  it('should render the component and matches it against stored snapshot in hindi', () => {
    const { asFragment } = setup('Hindi');

    expect(asFragment()).toMatchSnapshot();
  });

  it('should test the textarea text with valid language', async () => {
    setup('Hindi');

    userEvent.type(screen.getByRole('textbox', { name: 'addText (Hindi)' }), 'बपपप');

    await waitFor(() => {
      expect(screen.queryByText('typeInChosenLanguage')).not.toBeInTheDocument();
    });
  });

  it('should test the textarea text from virtual keyboard for "Hindi" language', async () => {
    setup('Hindi');
    userEvent.click(screen.getByRole('button', { name: 'keyboardBtn' }));

    await waitFor(() => {
      expect(screen.getByTestId('virtual-keyboard')).toBeInTheDocument();
    });

    userEvent.click(screen.getByText('caps'));
    userEvent.click(screen.getByText('क्ष'));

    expect(screen.getByRole('textbox', { name: 'addText (Hindi)' })).toHaveValue('क्ष');
  });

  it('should hide the virtual keyboard when user empty the textfield', async () => {
    setup('Hindi');

    userEvent.click(screen.getByRole('button', { name: 'keyboardBtn' }));

    await waitFor(() => {
      expect(screen.getByTestId('virtual-keyboard')).toBeInTheDocument();
    });

    userEvent.type(screen.getByRole('textbox', { name: 'addText (Hindi)' }), 'बपपप');

    userEvent.clear(screen.getByRole('textbox', { name: 'addText (Hindi)' }));

    await waitFor(() => {
      expect(screen.queryByText('X')).not.toBeInTheDocument();
    });
  });
});
