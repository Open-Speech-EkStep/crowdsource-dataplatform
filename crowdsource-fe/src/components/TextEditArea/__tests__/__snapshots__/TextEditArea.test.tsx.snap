// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditArea English should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    class="
          position-relative border border-2 border-primary
          addText false
          false rounded-8  bg-light p-4
        "
    data-testid="TextEditArea"
  >
    <div
      class="false"
    >
      <div>
        <label
          class="textareaLabel display-6 form-label"
          for="textarea"
        >
          addText (english)
        </label>
        <textarea
          class="textarea border-0 p-0 display-3 form-control"
          id="textarea"
          placeholder="typingPlaceholder"
        />
      </div>
      <div
        class="keyboardIcon d-none  d-md-block  position-absolute"
      >
        <button
          class="d-flex info normal btn btn-link"
          data-testid="Button"
          type="button"
        >
          <div
            style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
          >
            <div
              style="box-sizing: border-box; display: block; max-width: 100%;"
            >
              <img
                alt=""
                aria-hidden="true"
                src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjQiIGhlaWdodD0iMjQiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgdmVyc2lvbj0iMS4xIi8+"
                style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
              />
            </div>
            <img
              alt="keyboardBtn"
              data-nimg="intrinsic"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
